src/app/shared/document/document-detail/document-detail.component.ts

```diff
@@ -0,0 +1,401 @@
+import {selectDocumentById} from "../../../core/store/documents/documents.state";
+import {DocumentsAction} from "../../../core/store/documents/documents.action";
+import {IntervalObservable} from "rxjs/observable/IntervalObservable";
+import {selectCollectionById} from "../../../core/store/collections/collections.state";
+import {selectUserNameById} from "../../../core/store/users/users.state";
+import {filter, take} from "rxjs/operators";
+import {isNullOrUndefined} from "util";
+import {UsersAction} from "../../../core/store/users/users.action";
+import {selectOrganizationByWorkspace} from "../../../core/store/organizations/organizations.state";
+import {Observable} from "rxjs/Observable";
+import {CorrelationIdGenerator} from "../../../core/store/correlation-id.generator";
+import {CollectionsAction} from "../../../core/store/collections/collections.action";
+import {DetailRow} from "../detail-row";
+
+@Component({
+  selector: 'document-detail',
+  templateUrl: './document-detail.component.html',
+  styleUrls: ['./document-detail.component.scss']
+})
+export class DocumentDetailComponent implements OnInit, OnDestroy {
+
+  @Input()
+  public collection: CollectionModel;
+
+  public _documentModel: DocumentModel;
+
+  @Output()
+  public documentUpdate = new EventEmitter<DocumentModel>();
+
+  public rows: DetailRow[] = [];
+
+  public createdBy$: Observable<string>;
+  public updatedBy$: Observable<string>;
+
+  private subscriptions = new Subscription();
+
+  public summary: string;
+
+  constructor(private i18n: I18n,
+              private store: Store<AppState>,
+              private notificationService: NotificationService) { }
+
+  get documentModel(): DocumentModel {
+    return this._documentModel;
+  }
+
+  @Input("document")
+  set documentModel(model: DocumentModel) {
+    this._documentModel = model;
+    this.rows = [];
+    this.encodeDocument();
+  }
+
+  public ngOnInit() {
+    this.subscriptions.add(this.store.select(selectDocumentById(this._documentModel.id)).subscribe(doc => {
+      this._documentModel = doc;
+      this.encodeDocument();
+    }));
+    this.subscriptions.add(this.store.select(selectCollectionById(this.collection.id)).subscribe(col => {
+      this.collection = col;
+      this.encodeDocument();
+    }));
+    this.subscriptions.add(IntervalObservable.create(60000).subscribe(() => this.patchDocument()));
+
+    this.createdBy$ = this.store.select(selectUserNameById(this._documentModel.createdBy))
+      .pipe(filter(name => !isNullOrUndefined(name)));
+    this.updatedBy$ = this.store.select(selectUserNameById(this._documentModel.updatedBy))
+      .pipe(filter(name => !isNullOrUndefined(name)));
+
+    this.subscriptions.add(this.store.select(selectOrganizationByWorkspace)
+      .pipe(filter(org => !isNullOrUndefined(org)), take(1))
+      .subscribe(org => this.store.dispatch(new UsersAction.Get({ organizationId: org.id} ))));
+
+    this.encodeDocument();
```

Comment:
Why is `encodeDocument()` method called so many times? What about having a single stream merging document and collection together and then calling this method once?

Reply:
It is a little bit tricky. Those entites can be changed separately. I can only imagine caching some last hash code of both and checking whether there is a need to refresh anything. However, real changes occur only when there are changes to any of these structures, otherwise, the rows object remains the same so no unnecessary changes occur.